In [1]:
# Importing the python libraries for model.
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.offline as py
import plotly.graph_objs as go
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, confusion_matrix, log_loss

# Surpass the warnings.....
import warnings
warnings.filterwarnings('ignore')

# Step 1 -  Load the Data

In [2]:
# load data file.
data = pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv')

In [5]:
data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [4]:
data.shape

(1470, 35)

In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [35]:
data.dtypes.value_counts()

int64     26
object     9
dtype: int64

In [37]:
df_cat = pd.DataFrame(data = data.select_dtypes(include=['object']))
df_num = pd.DataFrame(data = data.select_dtypes(include=['int64']))
print("Categorical Columns", df_cat.shape)
print("Numerical Columns", df_num.shape)

Categorical Columns (1470, 9)
Numerical Columns (1470, 26)


**Observations:-**
- We have data of HR employees of the compnies.
- The dataframe has 1470 rows and 35 columns.
- There are total 9 columns has categorical in nature and 26 columns which has continus value in nature.

In [15]:
data.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count  1470.000000  ...               1470.000000         1470.0   
mean      2.063946  ...                  2.712245           80.0   
std       1.106940  ...                  1.081209            0.0   
min       1.000000  ...                  1.000000           80.0   
25%       1.000000  ...                  2.000000           80.0   
50%       2.000000  ...                  3.000000           80.0   
75%       3.000000  ...                  4.000000           80.0   
max       5.000000  ...                  4.000000           80.0   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count       1470.000000        1470.000000            1470.000000   
mean           0.793878          11.279592               2.799320   
std            0.852077           7.780782               1.289271   
min            0.000000           0.000000               0.000000   
25%            0.000000           6.000000               2.000000   
50%            1.000000          10.000000               3.000000   
75%            1.000000          15.000000               3.000000   
max            3.000000          40.000000               6.000000   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count      1470.000000     1470.000000         1470.000000   
mean          2.761224        7.008163            4.229252   
std           0.706476        6.126525            3.623137   
min           1.000000        0.000000            0.000000   
25%           2.000000        3.000000            2.000000   
50%           3.000000        5.000000            3.000000   
75%           3.000000        9.000000            7.000000   
max           4.000000       40.000000           18.000000   

       YearsSinceLastPromotion  YearsWithCurrManager  
count              1470.000000           1470.000000  
mean                  2.187755              4.123129  
std                   3.222430              3.568136  
min                   0.000000              0.000000  
25%                   0.000000              2.000000  
50%                   1.000000              3.000000  
75%                   3.000000              7.000000  
max                  15.000000             17.000000  

[8 rows x 26 columns]

#### Let's check the null values present in the data.

In [24]:
# Let's check is there is any column which have all values as null.
data.isnull().all(axis = 0).sum()

0

In [25]:
# Let's check is there is any row which have all values as null.
data.isnull().all(axis = 1).sum()

0

In [26]:
# Let's check is there is any cell which have values as null.
data.isnull().any(axis = 0).sum()

0

**Observations:-**
- There are no any column present in the data which have all values as null value.
- There are no any row present in the data which have all values as null value.
- There no any null value present in the data.

# Step 2 - Feature engineering

There are some column present in or data which are categorical in nature. For model building we have to convert these categorical values as nummerical form. For this we are using the dummie value creation.

There are some column present in or data which are numerical in nature but the range of the values are roo high. For model building we have to convert these columns in standarise form. For this we are using the Standare scaler for data scalling.

#### Dummie column creation

In [38]:
df_cat

Attrition     BusinessTravel              Department EducationField  \
0          Yes      Travel_Rarely                   Sales  Life Sciences   
1           No  Travel_Frequently  Research & Development  Life Sciences   
2          Yes      Travel_Rarely  Research & Development          Other   
3           No  Travel_Frequently  Research & Development  Life Sciences   
4           No      Travel_Rarely  Research & Development        Medical   
...        ...                ...                     ...            ...   
1465        No  Travel_Frequently  Research & Development        Medical   
1466        No      Travel_Rarely  Research & Development        Medical   
1467        No      Travel_Rarely  Research & Development  Life Sciences   
1468        No  Travel_Frequently                   Sales        Medical   
1469        No      Travel_Rarely  Research & Development        Medical   

      Gender                    JobRole MaritalStatus Over18 OverTime  
0     Female            Sales Executive        Single      Y      Yes  
1       Male         Research Scientist       Married      Y       No  
2       Male      Laboratory Technician        Single      Y      Yes  
3     Female         Research Scientist       Married      Y      Yes  
4       Male      Laboratory Technician       Married      Y       No  
...      ...                        ...           ...    ...      ...  
1465    Male      Laboratory Technician       Married      Y       No  
1466    Male  Healthcare Representative       Married      Y       No  
1467    Male     Manufacturing Director       Married      Y      Yes  
1468    Male            Sales Executive       Married      Y       No  
1469    Male      Laboratory Technician       Married      Y       No  

[1470 rows x 9 columns]

Here we have a feature 'Attrition' who is our target variable. So we have to exclude this feature from he categorical_cols list before creation of dummie values.

In [43]:
#### Create Dummie values ........
df_cat.drop('Attrition', axis = 1, inplace=True)

In [44]:
df_cat

BusinessTravel              Department EducationField  Gender  \
0         Travel_Rarely                   Sales  Life Sciences  Female   
1     Travel_Frequently  Research & Development  Life Sciences    Male   
2         Travel_Rarely  Research & Development          Other    Male   
3     Travel_Frequently  Research & Development  Life Sciences  Female   
4         Travel_Rarely  Research & Development        Medical    Male   
...                 ...                     ...            ...     ...   
1465  Travel_Frequently  Research & Development        Medical    Male   
1466      Travel_Rarely  Research & Development        Medical    Male   
1467      Travel_Rarely  Research & Development  Life Sciences    Male   
1468  Travel_Frequently                   Sales        Medical    Male   
1469      Travel_Rarely  Research & Development        Medical    Male   

                        JobRole MaritalStatus Over18 OverTime  
0               Sales Executive        Single      Y      Yes  
1            Research Scientist       Married      Y       No  
2         Laboratory Technician        Single      Y      Yes  
3            Research Scientist       Married      Y      Yes  
4         Laboratory Technician       Married      Y       No  
...                         ...           ...    ...      ...  
1465      Laboratory Technician       Married      Y       No  
1466  Healthcare Representative       Married      Y       No  
1467     Manufacturing Director       Married      Y      Yes  
1468            Sales Executive       Married      Y       No  
1469      Laboratory Technician       Married      Y       No  

[1470 rows x 8 columns]

In [50]:
for col in df_cat.columns:
    print("Column Name -> ", col)
    print(df_cat[col].value_counts())
    print(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")

Column Name ->  BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: BusinessTravel, dtype: int64
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Column Name ->  Department
Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Column Name ->  EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: EducationField, dtype: int64
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Column Name ->  Gender
Male      882
Female    588
Name: Gender, dtype: int64
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Column Name ->  JobRole
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145


From the above results we can say that there are one column present in our data 'Over18' which have onlly one value.
So we are going to delete this column because this column dose not do any help  for model building

In [51]:
df_cat.drop('Over18', axis = 1, inplace=True)

In [52]:
df_cat

BusinessTravel              Department EducationField  Gender  \
0         Travel_Rarely                   Sales  Life Sciences  Female   
1     Travel_Frequently  Research & Development  Life Sciences    Male   
2         Travel_Rarely  Research & Development          Other    Male   
3     Travel_Frequently  Research & Development  Life Sciences  Female   
4         Travel_Rarely  Research & Development        Medical    Male   
...                 ...                     ...            ...     ...   
1465  Travel_Frequently  Research & Development        Medical    Male   
1466      Travel_Rarely  Research & Development        Medical    Male   
1467      Travel_Rarely  Research & Development  Life Sciences    Male   
1468  Travel_Frequently                   Sales        Medical    Male   
1469      Travel_Rarely  Research & Development        Medical    Male   

                        JobRole MaritalStatus OverTime  
0               Sales Executive        Single      Yes  
1            Research Scientist       Married       No  
2         Laboratory Technician        Single      Yes  
3            Research Scientist       Married      Yes  
4         Laboratory Technician       Married       No  
...                         ...           ...      ...  
1465      Laboratory Technician       Married       No  
1466  Healthcare Representative       Married       No  
1467     Manufacturing Director       Married      Yes  
1468            Sales Executive       Married       No  
1469      Laboratory Technician       Married       No  

[1470 rows x 7 columns]

Now we are going to create the dummie value. Here we are using drop_first = True because we don't column repetation.

In [57]:
dummie_df = pd.get_dummies(df_cat, drop_first=True)

In [58]:
dummie_df

BusinessTravel_Travel_Frequently  BusinessTravel_Travel_Rarely  \
0                                    0                             1   
1                                    1                             0   
2                                    0                             1   
3                                    1                             0   
4                                    0                             1   
...                                ...                           ...   
1465                                 1                             0   
1466                                 0                             1   
1467                                 0                             1   
1468                                 1                             0   
1469                                 0                             1   

      Department_Research & Development  Department_Sales  \
0                                     0                 1   
1                                     1                 0   
2                                     1                 0   
3                                     1                 0   
4                                     1                 0   
...                                 ...               ...   
1465                                  1                 0   
1466                                  1                 0   
1467                                  1                 0   
1468                                  0                 1   
1469                                  1                 0   

      EducationField_Life Sciences  EducationField_Marketing  \
0                                1                         0   
1                                1                         0   
2                                0                         0   
3                                1                         0   
4                                0                         0   
...                            ...                       ...   
1465                             0                         0   
1466                             0                         0   
1467                             1                         0   
1468                             0                         0   
1469                             0                         0   

      EducationField_Medical  EducationField_Other  \
0                          0                     0   
1                          0                     0   
2                          0                     1   
3                          0                     0   
4                          1                     0   
...                      ...                   ...   
1465                       1                     0   
1466                       1                     0   
1467                       0                     0   
1468                       1                     0   
1469                       1                     0   

      EducationField_Technical Degree  Gender_Male  ...  \
0                                   0            0  ...   
1                                   0            1  ...   
2                                   0            1  ...   
3                                   0            0  ...   
4                                   0            1  ...   
...                               ...          ...  ...   
1465                                0            1  ...   
1466                                0            1  ...   
1467                                0            1  ...   
1468                                0            1  ...   
1469                                0            1  ...   

      JobRole_Laboratory Technician  JobRole_Manager  \
0                                 0                0   
1                                 0                0   
2                                 1                0   
3                                 0                0   
4

**Observations:-**
- After dummie variable creation we have 21 columns available in our dummie data.

#### Scaling the Numerical Columns

In [61]:
df_num

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
0      41       1102                 1          2              1   
1      49        279                 8          1              1   
2      37       1373                 2          2              1   
3      33       1392                 3          4              1   
4      27        591                 2          1              1   
...   ...        ...               ...        ...            ...   
1465   36        884                23          2              1   
1466   39        613                 6          1              1   
1467   27        155                 4          3              1   
1468   49       1023                 2          3              1   
1469   34        628                 8          3              1   

      EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
0                  1                        2          94               3   
1                  2                        3          61               2   
2                  4                        4          92               2   
3                  5                        4          56               3   
4                  7                        1          40               3   
...              ...                      ...         ...             ...   
1465            2061                        3          41               4   
1466            2062                        4          42               2   
1467            2064                        2          87               4   
1468            2065                        4          63               2   
1469            2068                        2          82               4   

      JobLevel  ...  RelationshipSatisfaction  StandardHours  \
0            2  ...                         1             80   
1            2  ...                         4             80   
2            1  ...                         2             80   
3            1  ...                         3             80   
4            1  ...                         4             80   
...        ...  ...                       ...            ...   
1465         2  ...                         3             80   
1466         3  ...                         1             80   
1467         2  ...                         2             80   
1468         2  ...                         4             80   
1469         2  ...                         1             80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

      WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                   1               6                   4   
1                   3              10                   7   
2                   3               0                   0   
3                   3               8                   7   
4                   3               2                   2   
...               ...             ...                 ...   
1465                3               5                   2   
1466                3               7                   7   
1467                3               6                   2   
1468                2               9                   6   
1469   

In [62]:
# For this we are using standard scaler.
sc = StandardScaler()

In [64]:
sc_df = sc.fit_transform(df_num)

In [66]:
df_num_scaled = pd.DataFrame(data = sc_df, columns = df_num.columns, index = df_num.index)

In [68]:
df_num_scaled.describe()

Age     DailyRate  DistanceFromHome     Education  \
count  1.470000e+03  1.470000e+03      1.470000e+03  1.470000e+03   
mean  -4.017950e-17  7.197568e-17      5.164048e-17  2.697011e-16   
std    1.000340e+00  1.000340e+00      1.000340e+00  1.000340e+00   
min   -2.072192e+00 -1.736576e+00     -1.010909e+00 -1.868426e+00   
25%   -7.581700e-01 -8.366616e-01     -8.875151e-01 -8.916883e-01   
50%   -1.011589e-01 -1.204135e-03     -2.705440e-01  8.504925e-02   
75%    6.653541e-01  8.788772e-01      5.932157e-01  1.061787e+00   
max    2.526886e+00  1.726730e+00      2.444129e+00  2.038524e+00   

       EmployeeCount  EmployeeNumber  EnvironmentSatisfaction    HourlyRate  \
count         1470.0    1.470000e+03             1.470000e+03  1.470000e+03   
mean             0.0    1.132881e-16             7.054070e-17  1.728021e-16   
std              0.0    1.000340e+00             1.000340e+00  1.000340e+00   
min              0.0   -1.701283e+00            -1.575686e+00 -1.766079e+00   
25%              0.0   -8.866700e-01            -6.605307e-01 -8.803615e-01   
50%              0.0   -7.253514e-03             2.546249e-01  5.355811e-03   
75%              0.0    8.821327e-01             1.169781e+00  8.787715e-01   
max              0.0    1.733302e+00             1.169781e+00  1.678377e+00   

       JobInvolvement      JobLevel  ...  RelationshipSatisfaction  \
count    1.470000e+03  1.470000e+03  ...              1.470000e+03   
mean    -6.495182e-18  1.479542e-16  ...             -1.910792e-16   
std      1.000340e+00  1.000340e+00  ...              1.000340e+00   
min     -2.432006e+00 -9.614864e-01  ...             -1.584178e+00   
25%     -1.026167e+00 -9.614864e-01  ...             -6.589728e-01   
50%      3.796721e-01 -5.778755e-02  ...              2.662326e-01   
75%      3.796721e-01  8.459113e-01  ...              1.191438e+00   
max      1.785511e+00  2.653309e+00  ...              1.191438e+00   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count         1470.0      1.470000e+03       1.470000e+03   
mean             0.0      6.600918e-17      -9.236753e-17   
std              0.0      1.000340e+00       1.000340e+00   
min              0.0     -9.320144e-01      -1.450167e+00   
25%              0.0     -9.320144e-01      -6.787735e-01   
50%              0.0      2.419883e-01      -1.645114e-01   
75%              0.0      2.419883e-01       4.783162e-01   
max              0.0      2.589994e+00       3.692454e+00   

       TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
count           1.470000e+03     1.470000e+03    1.470000e+03   
mean           -1.106069e-16    -5.165936e-17   -1.941002e-17   
std             1.000340e+00     1.000340e+00    1.000340e+00   
min            -2.171982e+00    -2.493820e+00   -1.144294e+00   
25%            -6.201892e-01    -1.077862e+00   -6.544537e-01   
50%             1.557071e-01     3.380962e-01   -3.278933e-01   
75%             1.557071e-01     3.380962e-01    3.252275e-01   
max             2.483396e+00     1.754054e+00    5.386914e+00   

       YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
count        1.470000e+03             1.470000e+03          1.470000e+03  
mean         6.117555e-17            -4.236974e-17         -5.823006e-17  
std          1.000340e+00             1.000340e+00          1.000340e+00  
min         -1.167687e+00            -6.791457e-01         -1.155935e+00  
25%         -6.154916e-01            -6.791457e-01         -5.952272e-01  
50%         -3.393937e-01            -3.687153e-01         -3.148735e-01  
75%          7.649976e-01             2.521455e-01          8.065415e-01  
max          3.802074e+00             3.977310e+00          3.610079e+00  

[8 rows x 26 columns]

In [69]:
# let's create our final_data.
df_final = pd.concat([df_num_scaled, dummie_df], axis=1)

In [70]:
df_final.shape

(1470, 47)

In [76]:
target = data['Attrition']
target

0       Yes
1        No
2       Yes
3        No
4        No
       ... 
1465     No
1466     No
1467     No
1468     No
1469     No
Name: Attrition, Length: 1470, dtype: object

In [77]:
# maopping the Yes = 1 and No = 0.
map = {'Yes' : 1, 'No' : 0}
target = target.apply(lambda x : map[x])
target.shape

(1470,)

In [78]:
target

0       1
1       0
2       1
3       0
4       0
       ..
1465    0
1466    0
1467    0
1468    0
1469    0
Name: Attrition, Length: 1470, dtype: int64

In [79]:
X = df_final
y = target